### PR TITLE
Ready For Review

### PR DESCRIPTION
**Pull request by JorgeOehrens**
React-based decentralized application (dApp) that interacts with a smart contract on the Stellar blockchain. It allows users to manage and update their "Soroban Vitae" (CV) by submitting personal information such as name, email, skills, experience, and education.

- [Create smart contract and test](https://github.com/paltalabs/create-soroban-dapp/commit/b7078fbc5a5700ba9823890ab42c1cda08262e0c)
- [Added auth for smart contract, create new test and update cargo](https://github.com/paltalabs/create-soroban-dapp/commit/57765df683b13e49a3db04668aa345976fe5ef3f)
- [Modify UI in Deapp and connect with smart contact](https://github.com/paltalabs/create-soroban-dapp/commit/11d02a32eddbd9cac36767eeb2fcc25aa0edb287)
- [Update README.md](https://github.com/paltalabs/create-soroban-dapp/commit/554871f5f1558c549189ca98b66b726c0e06585b)
- [Merge pull request](https://github.com/paltalabs/create-soroban-dapp/commit/eae154d9d445bf4f585f57c15af837e95f558664)
- [Deploy and fixed cargo cv](https://github.com/paltalabs/create-soroban-dapp/commit/bd165023b7d76ab06d4e98a3ca7d9f3e702f2134)
- [Functional UI and Auth for smart contracts](https://github.com/paltalabs/create-soroban-dapp/commit/0c0e6bf2cff0118c95655e1c8de60f47bca02ed1)
- [Fixed adress for read the last soroban vitae](https://github.com/paltalabs/create-soroban-dapp/commit/4cd5bfb6657771acc62fcb77217927c0c650f0c5)
- [Update README.md](https://github.com/paltalabs/create-soroban-dapp/commit/946718dba5d25056a03352e04a27e175caaef475)
- [Merge pull request](https://github.com/paltalabs/create-soroban-dapp/commit/f97ec6caa94846080c2a3c26dae60c99437838f4)
- [Merge pull request](https://github.com/paltalabs/create-soroban-dapp/commit/ca8b57c2a97c8ea3087825d033c1549e772a3ef1)
- [Added cv typed, fixed error in vercel](https://github.com/paltalabs/create-soroban-dapp/commit/82ea246efdbc774027d84b39ab81306f720db228)
- [Merge pull request](https://github.com/paltalabs/create-soroban-dapp/commit/b8cc21d33c8f948c92bfba5a991bc2335ea4de72)

### CODE REVIEW

**Comment by esteblock** [comment]
Vercel preview: https://create-soroban-dapp-git-fork-jorgeoehrens-feat-43a886-paltalabs.vercel.app/

File: soroban-react-dapp/contracts/cv/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    experience: Vec<String>,
+    education: Vec<String>,
+}
+
+#[contract]
+pub struct CVContract;
+
+#[contractimpl]
+impl CVContract {
+    pub fn init(env: Env, owner: Address) {
```
Comment (yripper) [inline soroban-react-dapp/contracts/cv/src/lib.rs:19]: The `init` function is currently overwriting any existing data for the `owner` in the contract's storage. This introduces a significant security risk, as it allows any user to invoke this function and delete or replace data for any other user. 

Moreover, the `init` function seems unnecessary for the intended functionality of this contract. Initialization logic should be carefully managed to avoid unintended side effects like this.

File: soroban-react-dapp/contracts/cv/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        cv.experience.push_back(experience);
+        env.storage().instance().set(&owner, &cv);
+    }
+
+    pub fn add_education(env: Env, owner: Address, education: String) {
+        require_auth(&env, &owner);
+        let mut cv = env.storage().instance().get::<Address, CV>(&owner).unwrap();
+        cv.education.push_back(education);
+        env.storage().instance().set(&owner, &cv);
+    }
```
Comment (yripper) [inline soroban-react-dapp/contracts/cv/src/lib.rs:57]: The implementation looks solid, but there are some potential optimizations and security considerations:

1. **Transaction Efficiency:** Currently, the user has to sign a transaction for each function (add_skill, add_experience, add_education), which is not ideal. Consolidating these into a single update_cv method that can update all fields at once would be more efficient and user-friendly.
2. **Security Concern:** While the use of require_auth is appropriate, the presence of the init function poses a security risk, as it allows anyone to overwrite existing data. This needs to be addressed to prevent unauthorized data modification.
3. **Error Handling:** The update_cv method could be enhanced by validating the input fields. If certain fields are not provided, the function should throw an error to ensure data integrity.

File: soroban-react-dapp/contracts/cv/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    pub fn get_cv(env: Env, owner: Address) -> String {
+        let cv = env.storage().instance().get(&owner).unwrap();
+        cv.name.to_string() + " " + cv.email.to_string() + " " + cv.skills.to_string() + " " + cv.experience.to_string() + " " + cv.education.to_string()
+    }
```
Comment (yripper) [inline soroban-react-dapp/contracts/cv/src/lib.rs:62]: This method would be more effective if it returned the `CV` struct instead of a concatenated string. Additionally, using `unwrap()` could cause a panic if the key doesn't exist. Instead, consider using `unwrap_or()` to safely return an empty `CV` if the key is not found:

```rust
pub fn get_cv(env: Env, owner: Address) -> CV {
    let empty_cv = CV {
        owner: owner.clone(),
        name: String::from_str(&env, ""),
        email: String::from_str(&env, ""),
        skills: Vec::new(&env),
        experience: Vec::new(&env),
        education: Vec::new(&env),
    };
    let cv = env.storage().instance().get(&owner).unwrap_or(empty_cv);
    cv
}
```

This approach prevents panics and ensures the function always returns a valid `CV` struct.

File: soroban-react-dapp/contracts/cv/src/test.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+#![cfg(test)]
+
+use super::*;
+use soroban_sdk::{testutils::Address as _, Address, Env, String, Vec};
+
+#[test]
+fn test_cv_contract() {
```
Comment (yripper) [inline soroban-react-dapp/contracts/cv/src/test.rs:7]: The current test only verifies the expected behavior and contains a single test case. To ensure robustness, you should add tests for each method individually, covering various edge cases. This includes:

1. **Multiple Initializations**: Test what happens when `init` is called multiple times for the same `owner`.
2. **Security Concerns**: Add tests to simulate potential attacks, such as unauthorized overwriting of another user’s data via the `init` method.
3. **Error Handling**: Ensure that edge cases, like missing fields or invalid data, are properly handled.

Expanding the test coverage in these areas will help ensure the contract is secure and behaves as expected under all scenarios.

**Comment by yripper** [comment]
The frontend could be enhanced by re-fetching the user's data after each contract method is executed. Additionally, there are too many transactions being signed; this could be consolidated into a single transaction. The `init` method is unnecessary for this contract and should be removed.

Furthermore, the instructions specified the addition of a whitelist to restrict access to the `set_title` method, ensuring only authorized users could invoke it. While the smart contract developed here is a good concept, it doesn't fully align with the requirements we outlined.
